pkg/cloudprovider/provider/alibaba/provider.go

```diff
@@ -0,0 +1,413 @@
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.AccessKeyID == "" {
+		return fmt.Errorf("accessKeyID is missing")
+	}
+	if c.AccessKeySecret == "" {
+		return fmt.Errorf("accessKeySecret is missing")
+	}
+	if c.RegionID == "" {
+		return fmt.Errorf("regionID is missing")
+	}
+	if c.InstanceType == "" {
+		return fmt.Errorf("instanceType is missing")
+	}
+	if c.VSwitchID == "" {
+		return fmt.Errorf("vSwitchID is missing")
+	}
+	if c.InternetMaxBandwidthOut == "" {
+		return fmt.Errorf("internetMaxBandwidthOut is missing")
+	}
+	if c.ZoneID == "" {
+		return fmt.Errorf("zoneID is missing")
+	}
+	_, err = getImageIDForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	return nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+
+	i, err := getInstance(client, machine.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if i != nil {
+		return &alibabaInstance{instance: i}, nil
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (config string, name string, err error) {
+	return "", "", nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.Labels == nil {
+		c.Labels = map[string]string{}
+	}
+
+	c.Labels[machineUIDTag] = string(machine.UID)
+
+	createInstanceRequest := ecs.CreateCreateInstanceRequest()
+	createInstanceRequest.ImageId, _ = getImageIDForOS(pc.OperatingSystem)
+	createInstanceRequest.InstanceName = machine.Name
+	createInstanceRequest.InstanceType = c.InstanceType
+	createInstanceRequest.VSwitchId = c.VSwitchID
+	createInstanceRequest.InternetMaxBandwidthOut = requests.Integer(c.InternetMaxBandwidthOut)
+	encodedUserData := base64.StdEncoding.EncodeToString([]byte(userdata))
+	createInstanceRequest.UserData = encodedUserData
+	createInstanceRequest.SystemDiskCategory = "cloud_efficiency"
+	createInstanceRequest.ZoneId = c.ZoneID
+
+	res, err := client.CreateInstance(createInstanceRequest)
```

Comment:
If the creation is a multi-step process, the provider needs to add a distinct finalizer for each step, to make sure we clean up even if the creation failed halfways. See the azure provider as an example

Reply:
The only thing that we allocate is a public IP and it gets deleted once the instance is deleted by default, thus I am not quite sure that it is necessary. especially that we don't invoke any deletion for the IP addresses in the cleanup method since its an implicit part of the instance cleanup.

Reply:
It definitely is necesary because if we successfully created the instance but got an error trying to allocate the IP, `Create` will return an error which results in the cleanup finalizer not being added which results in us orphaning the instance.

Reply:
I have added the finalizer, small note though, the allocating of the public IP address happens in the `Get` method, which means it will try to allocate the IP address multiple times.